### PR TITLE
Pen Allocation Grouping Algorithm Pull request

### PR DESCRIPTION
**Pull request by cvankerkhove**
This pull request contains updates to the animal management system; specifically a file is added (found in routines/animal) that contains a lactating cow pen grouping by ration requirements. Furthermore, there are other small modifications to animal_management and cows in order to implement the algorithm. (Note that this can be tested for any number of inputs for pens, but the more pens, the longer the simulation)

### CODE REVIEW

**Comment by militsasotirova** [comment]
I was able to run the simulation and everything looked fine, but could you add some documentation in the grouping() method in clustering_pen_grouping.py? I’m trying to follow what is happening, but I’m getting a little lost with the variable names and what they mean.

We should also change the installation instructions to include installing the “pandas”, “scipy”, and “xlsxwriter” modules.

**Review by KFosterReed — COMMENTED** [review]
Can you explain what all the pulp.sol and pulp.mps files are? Are they output functions? Are they part of a test? Just trying to figure out how I should be interpreting them. Thanks!

File: RUFAS/routines/animal/clustering_pen_grouping.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        x = percentileofscore(l, e)
+        percentile_list.append(x/100)
+    return(percentile_list)
+
+
+#############################################
+#Initial Data Manipulation
+#############################################
+
+def grouping(list):
```
Comment (KFosterReed) [inline RUFAS/routines/animal/clustering_pen_grouping.py]: I would like to add a formal function description to all functions.

File: RUFAS/routines/animal/clustering_pen_grouping.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+    ID = []
```
Comment (KFosterReed) [inline RUFAS/routines/animal/clustering_pen_grouping.py]: We need to define all of these variables

File: RUFAS/routines/animal/clustering_pen_grouping.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            if (n == weeknlist[i] and LACTcat[i] == 1):
+                rank.append(u)
+                u += 1
+            elif (n == weeknlist[i] and LACTcat[i] == 2):
+                rank.append(v)
+                v += 1
+    n = len(NGdata.columns)
+    NGdata.insert(n, 'rank', rank)
+
+    ##Size per pen category##
```
Comment (KFosterReed) [inline RUFAS/routines/animal/clustering_pen_grouping.py:138]: These look like hard-coded values for the size of pens? If they are, we should change them to be pen attributes

Comment (KFosterReed) [reply RUFAS/routines/animal/clustering_pen_grouping.py:138]: If they aren't pen sizes... then what are they?

File: RUFAS/routines/animal/clustering_pen_grouping.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    SFPenP=148 #Size Fresh Pens Pimiparous
+    SFPenM=148 #Size Fresh Pens Multiparous
+    SEPenP=148 #Size Early Pens Pimiparous
+    SEPenM=148 #Size Early Pens Multiparous
+    SPPenP=444 #Size Peak Pens Pimiparous
+    SPPenM=444 #Size Peak Pens Multiparous
+
+    #Adding column DIMcat (Days in Milk Categorization)
+    ## 1=Fresh, 2=Early, 3=Pick, 4=Late##
```
Comment (KFosterReed) [inline RUFAS/routines/animal/clustering_pen_grouping.py]:
```suggestion
    ## 1=Fresh, 2=Early, 3=Peak, 4=Late##
```

File: RUFAS/routines/animal/clustering_pen_grouping.py (hunk 3 above)
Comment (cvankerkhove) [reply RUFAS/routines/animal/clustering_pen_grouping.py:138]: Yes, those are hard-coded values; they are not sizes for the actual pens themselves, but rather values to measure the rank given to the cow and categorize them correctly. Regardless I agree they probably should not be hard-coded but those numbers were taken directly from Jorge's R code and I'm not exactly sure where we would pull those values from. I have reached out to him about setting up a meeting this week to work out a solution to this. Thanks for bringing this to my attention

**Review by KFosterReed — CHANGES_REQUESTED** [review]
SImplify algorithm as discussed

**Pull request merged**